File: src.cpp/alib.unittests/ut_alib_strings_tokenizer.cpp
// #################################################################################################
//  aworx - Unit Tests
//
//  (c) 2013-2016 A-Worx GmbH, Germany
//  Published under MIT License (Open Source License, see LICENSE.txt)
// #################################################################################################
#include "alib/stdafx_alib.h"

#include "alib/strings/tokenizer.hpp"
#include "alib/compatibility/std_iostream.hpp"

#define TESTCLASSNAME       CPP_ALib_StringsTokenizer
#include "aworx_unittests.hpp"

#include <iostream>
#include <iomanip>

using namespace std;
using namespace aworx;

namespace ut_aworx {



UT_CLASS()



//--------------------------------------------------------------------------------------------------
//--- Tokenizer
//--------------------------------------------------------------------------------------------------
void tokenizerTest( const char* inputString, AString& res, char delim, char newDelim,
                    Whitespaces trim, int inpStart= -1, int inpEnd= -1  )
{
    Substring inp( inputString );
    if ( inpStart < 0 )  inpStart= 0;
    if ( inpEnd   < 0 )  inpEnd=   inp.Length() - 1;
    inp.Set( inp, inpStart, inpEnd-inpStart +1 );

    res.Clear();

    Tokenizer tknzr( inp, delim );

    while( tknzr.HasNext() )
    {
        res._( tknzr.Next(trim) );
        res._( newDelim );
    }

}

UT_METHOD( Tokenize )
{
    UT_INIT();

    AString as;
    AString res;

    // tokenizing empty string
    as.Clear()._( "" );
    res.Clear();
    {
        Tokenizer tknzr( as, ',' );  UT_EQ( true,  tknzr.HasNext() );
        res._( tknzr.Next() );                UT_EQ( "", res );
    }

    // tokenizing no delim
    as.Clear()._( "abc" );
    res.Clear();
    {
        Tokenizer tknzr( as, ',' );           UT_EQ( true,  tknzr.HasNext() );
        res._( tknzr.Next() );                UT_EQ( "abc", res );
    }

    // tokenizing
    {
        tokenizerTest( "abc",                   res, ',', '@',   Whitespaces::Trim,  -1,  -1 );   UT_EQ( res, "abc@"                );
        tokenizerTest( "a,bc",                  res, ',', '@',   Whitespaces::Trim,  -1,  -1 );   UT_EQ( res, "a@bc@"               );
        tokenizerTest( "a,bc",                  res, ',', '@',   Whitespaces::Trim,  -1,  -1 );   UT_EQ( res, "a@bc@"               );
        tokenizerTest( ",",                     res, ',', '@',   Whitespaces::Trim,  -1,  -1 );   UT_EQ( res, "@@"                  );
        tokenizerTest( ",,",                    res, ',', '@',   Whitespaces::Trim,  -1,  -1 );   UT_EQ( res, "@@@"                 );
        tokenizerTest( "a,b,c,,",               res, ',', '@',   Whitespaces::Trim,  -1,  -1 );   UT_EQ( res, "a@b@c@@@"            );
        tokenizerTest( "a,b,c",                 res, ',', '@',   Whitespaces::Trim,  -1,  -1 );   UT_EQ( res, "a@b@c@"              );
        tokenizerTest( ",a,b,c",                res, ',', '@',   Whitespaces::Trim,  -1,  -1 );   UT_EQ( res, "@a@b@c@"             );
        tokenizerTest( "123567",                res, ',', '@',   Whitespaces::Trim,   2,   2 );   UT_EQ( res, "3@"                  );
        tokenizerTest( "123567",                res, ',', '@',   Whitespaces::Trim,   2,   1 );   UT_EQ( res, "@"                   );
        tokenizerTest( "  abc , def , ghe ",    res, ',', '@',   Whitespaces::Trim,  -1,  -1 );   UT_EQ( res, "abc@def@ghe@"        );
        tokenizerTest( "abc , def,ghe,",        res, ',', '@',   Whitespaces::Trim,  -1,  -1 );   UT_EQ( res, "abc@def@ghe@@"       );
        tokenizerTest( "  abc , def , ghe ",    res, ',', '@',   Whitespaces::Keep,  -1,  -1 );   UT_EQ( res, "  abc @ def @ ghe @" );
        tokenizerTest( "abc , def,ghe,",        res, ',', '@',   Whitespaces::Keep,  -1,  -1 );   UT_EQ( res, "abc @ def@ghe@@"     );
    }

    // tokenizing with different delimiters
    {
        as.Clear()._( "1,5;3@4" );
        Tokenizer tknzr(as, ',');
        res= tknzr.Next();                            UT_EQ( "1",   res );  UT_EQ( true,  tknzr.HasNext() );
        res= tknzr.Next( Whitespaces::Trim, ';' );    UT_EQ( "5",   res );  UT_EQ( true,  tknzr.HasNext() );
        res= tknzr.Next( Whitespaces::Trim, '@' );    UT_EQ( "3",   res );  UT_EQ( true,  tknzr.HasNext() );
        res= tknzr.Next( Whitespaces::Trim, '-' );    UT_EQ( "4",   res );  UT_EQ( false, tknzr.HasNext() );
    }

    // tokenizing with different delimiters
    {
        as.Clear()._( "abc, 5;\t3;;; 4  " );
        Tokenizer tknzr(as,',');
        res= tknzr.Next();                          UT_EQ( "abc", res );  UT_EQ( true,  tknzr.HasNext() );
        res= tknzr.Next( Whitespaces::Trim, ';' );  UT_EQ( "5",   res );  UT_EQ( true,  tknzr.HasNext() );
        res= tknzr.Next();                          UT_EQ( "3",   res );  UT_EQ( true,  tknzr.HasNext() );
        res= tknzr.Next();                          UT_EQ( "",    res );  UT_EQ( true,  tknzr.HasNext() );
        res= tknzr.GetRest();                       UT_EQ( "; 4", res );  UT_EQ( false, tknzr.HasNext() );
    }

    // sub-tokens
    {
        as.Clear()._( "1,2;3 , 4;5,;," );
        Tokenizer tknzr(as, ';');

        Tokenizer tknzr2( tknzr.Next(), ',');
        res= tknzr2.Next(); UT_EQ( "1", res );  UT_TRUE(  tknzr2.HasNext() );
        res= tknzr2.Next(); UT_EQ( "2", res );  UT_TRUE( !tknzr2.HasNext() );
        UT_TRUE( tknzr.HasNext() );

        tknzr2.Set( tknzr.Next(), ',');
        res= tknzr2.Next(); UT_EQ( "3", res );  UT_TRUE(  tknzr2.HasNext() );
        res= tknzr2.Next(); UT_EQ( "4", res );  UT_TRUE( !tknzr2.HasNext() );
        UT_TRUE( tknzr.HasNext() );

        tknzr2.Set( tknzr.Next(), ',');
        res= tknzr2.Next(); UT_EQ( "5", res );  UT_TRUE(  tknzr2.HasNext() );
        res= tknzr2.Next(); UT_EQ( "",  res );  UT_TRUE( !tknzr2.HasNext() );
        UT_TRUE( tknzr.HasNext() );

        tknzr2.Set( tknzr.Next(), ',');
        res= tknzr2.Next(); UT_EQ( "", res );  UT_TRUE(  tknzr2.HasNext() );
        res= tknzr2.Next(); UT_EQ( "", res );  UT_TRUE( !tknzr2.HasNext() );
        UT_TRUE( !tknzr.HasNext() );
    }
}



UT_CLASS_END

} //namespace

File: src.java/alib.unittests/ut_com_alib/UT_alib_strings_Tokenizer.java
// #################################################################################################
//  Unit Tests - AWorx Library
//
//  (c) 2013-2016 A-Worx GmbH, Germany
//  Published under MIT License (Open Source License, see LICENSE.txt)
// #################################################################################################
package ut_com_alib;
import org.junit.Test;

import ut_com_aworx_uttools.UnitTest;
import ut_com_aworx_uttools.UnitTestSampleWriter;

import com.aworx.lib.enums.Whitespaces;
import com.aworx.lib.strings.AString;
import com.aworx.lib.strings.Substring;
import com.aworx.lib.strings.Tokenizer;


public class UT_alib_strings_Tokenizer extends UnitTest
{

//##################################################################################################
// SAMPLE code of class documentation
//##################################################################################################

void documentationSampleTokenizer()
{
    UnitTestSampleWriter utsw= new UnitTestSampleWriter("DOC_SAMPLES_ALIB_LIB_STRING_TOKEN.txt");

//! [DOC_SAMPLES_ALIB_LIB_STRING_TOKEN]
    // data string to tokenize
    AString data= new AString( "test;  abc ; 1,2 , 3 ; xyz ; including;separator" );

    // create tokenizer on data with ';' as delimiter
    Tokenizer tokens= new Tokenizer( data, ';' );

    // read tokens
    System.out.println( tokens.next().toString() ); // will print "test"
    System.out.println( tokens.next().toString() ); // will print "abc"
    System.out.println( tokens.next().toString() ); // will print "1,2 , 3"

    // tokenize actual (third) token (nested tokenizer)
    Tokenizer subTokens= new Tokenizer( tokens.actual,  ',');
    System.out.print( subTokens.next().toString() );

    while( subTokens.hasNext() )
        System.out.print( "~" + subTokens.next().toString() );

    System.out.println();

    // continue with the main tokenizer
    System.out.println( tokens.next().toString() ); // will print "xyz"

    // grab the rest, as we know that the last token might include our separator character
    System.out.println( tokens.getRest().toString() ); // will print "including;separator"
//! [DOC_SAMPLES_ALIB_LIB_STRING_TOKEN]

    utsw.flushAndResetSystemOut();
}

    
    @SuppressWarnings("static-method")
    @Test
    public void DocumentationSamples()
    {
    
        //UT_PRINT("*** Documentation Sample +**")
        documentationSampleTokenizer();
    }
    

    void tokenizerTest( String inputString, AString res, char delim, char newDelim,
                        Whitespaces trim, int inpStart, int inpEnd  )
    {
        Substring inp= new Substring( inputString );
        if ( inpStart >= 0 ) inp.start= inpStart;
        if ( inpEnd   >= 0 ) inp.end=   inpEnd;
        res.clear();

        Tokenizer tok= new Tokenizer( inp, delim );

        while( tok.hasNext() )
        {
            tok.next(trim).copyTo( res, true );
            res._( newDelim );
        }

    }

    void tokenizerTest( String inputString, AString res, char delim, char newDelim,
                        Whitespaces trim, int inpStart  )
    {
        tokenizerTest( inputString, res, delim, newDelim, trim, inpStart, -1);
    }

    void tokenizerTest( String inputString, AString res, char delim, char newDelim,
                        Whitespaces trim )
    {
        tokenizerTest( inputString, res, delim, newDelim, trim, -1, -1);
    }

    @SuppressWarnings("static-method")
    @Test
    public void Tokenize()
    {
        AString astr=  new AString();
        AString res=   new AString();

        // tokenizing empty string
        astr.clear()._( "" );
        res.clear();
        {
            Tokenizer tok= new Tokenizer( astr, ',' );   UT_EQ( true,  tok.hasNext() );
            tok.next().copyTo( res, true );              UT_EQ( "", res );
        }

        // tokenizing no delim
        astr.clear()._( "abc" );
        res.clear();
        {
            Tokenizer tok= new Tokenizer( astr, ',' );                UT_EQ( true,  tok.hasNext() );
            tok.next().copyTo( res, true );     UT_EQ( "abc", res );
        }

        // tokenizing
        {
            tokenizerTest( "abc",                res, ',', '@',   Whitespaces.TRIM, -1,   -1 );  UT_EQ ( "abc@",               res );
            tokenizerTest( "a,bc",               res, ',', '@',   Whitespaces.TRIM, -1,   -1 );  UT_EQ ( "a@bc@",              res );
            tokenizerTest( "a,bc",               res, ',', '@',   Whitespaces.TRIM, -1,   -1 );  UT_EQ ( "a@bc@",              res );
            tokenizerTest( ",",                  res, ',', '@',   Whitespaces.TRIM, -1,   -1 );  UT_EQ ( "@@",                 res );
            tokenizerTest( ",,",                 res, ',', '@',   Whitespaces.TRIM, -1,   -1 );  UT_EQ ( "@@@",                res );
            tokenizerTest( "a,b,c,,",            res, ',', '@',   Whitespaces.TRIM, -1,   -1 );  UT_EQ ( "a@b@c@@@",           res );
            tokenizerTest( "a,b,c",              res, ',', '@',   Whitespaces.TRIM, -1,   -1 );  UT_EQ ( "a@b@c@",             res );
            tokenizerTest( ",a,b,c",             res, ',', '@',   Whitespaces.TRIM, -1,   -1 );  UT_EQ ( "@a@b@c@",            res );
            tokenizerTest( "123567",             res, ',', '@',   Whitespaces.TRIM,  2,    2 );  UT_EQ ( "3@",                 res );
            tokenizerTest( "123567",             res, ',', '@',   Whitespaces.TRIM,  2,    1 );  UT_EQ ( "@",                  res );
            tokenizerTest( "  abc , def , ghe ", res, ',', '@',   Whitespaces.TRIM, -1,   -1 );  UT_EQ ( "abc@def@ghe@",       res );
            tokenizerTest( "abc , def,ghe,",     res, ',', '@',   Whitespaces.TRIM, -1,   -1 );  UT_EQ ( "abc@def@ghe@@",      res );
            tokenizerTest( "  abc , def , ghe ", res, ',', '@',   Whitespaces.KEEP,-1,   -1 );  UT_EQ ( "  abc @ def @ ghe @",res );
            tokenizerTest( "abc , def,ghe,",     res, ',', '@',   Whitespaces.KEEP,-1,   -1 );  UT_EQ ( "abc @ def@ghe@@",    res );
        }

        // tokenizing with different delimiters
        {
            astr.clear()._( "1,5;3@4" );
            Tokenizer tok= new Tokenizer(astr, ',');
            tok.next()                       .copyTo( res );    UT_EQ ( "1",   res );  UT_EQ( true,  tok.hasNext() );
            tok.next( Whitespaces.TRIM, ';' ).copyTo( res );    UT_EQ ( "5",   res );  UT_EQ( true,  tok.hasNext() );
            tok.next( Whitespaces.TRIM, '@' ).copyTo( res );    UT_EQ ( "3",   res );  UT_EQ( true,  tok.hasNext() );
            tok.next( Whitespaces.TRIM, '-' ).copyTo( res );    UT_EQ ( "4",   res );  UT_EQ( false, tok.hasNext() );
        }

        // tokenizing with different delimiters
        {
            astr.clear()._( "abc, 5;\t3;;; 4  " );
            Tokenizer tok= new Tokenizer(astr,',');
            tok.next()                       .copyTo( res );    UT_EQ ( "abc", res );  UT_EQ( true,  tok.hasNext() );
            tok.next( Whitespaces.TRIM, ';' ).copyTo( res );    UT_EQ ( "5",   res );  UT_EQ( true,  tok.hasNext() );
            tok.next()                       .copyTo( res );    UT_EQ ( "3",   res );  UT_EQ( true,  tok.hasNext() );
            tok.next()                        .copyTo( res );    UT_EQ ( "",    res );  UT_EQ( true,  tok.hasNext() );
            tok.getRest()                    .copyTo( res );    UT_EQ ( "; 4", res );  UT_EQ( false, tok.hasNext() );
        }

        // sub-tokens
        {
            astr.clear()._( "1,2;3 , 4;5,;," );
            Tokenizer tok= new Tokenizer(astr, ';');

            Tokenizer tok2= new Tokenizer( tok.next(), ',');
            tok2.next().copyTo( res ); UT_EQ ( "1", res );  UT_TRUE(  tok2.hasNext() );
            tok2.next().copyTo( res ); UT_EQ ( "2", res );  UT_TRUE( !tok2.hasNext() );
            UT_TRUE( tok.hasNext() );

            tok2.set( tok.next(), ',');
            tok2.next().copyTo( res ); UT_EQ ( "3", res );  UT_TRUE(  tok2.hasNext() );
            tok2.next().copyTo( res ); UT_EQ ( "4", res );  UT_TRUE( !tok2.hasNext() );
            UT_TRUE( tok.hasNext() );

            tok2.set( tok.next(), ',');
            tok2.next().copyTo( res ); UT_EQ ( "5", res );  UT_TRUE(  tok2.hasNext() );
            tok2.next().copyTo( res ); UT_EQ ( "",  res );  UT_TRUE( !tok2.hasNext() );
            UT_TRUE( tok.hasNext() );

            tok2.set( tok.next(), ',');
            tok2.next().copyTo( res ); UT_EQ ( "", res );  UT_TRUE(  tok2.hasNext() );
            tok2.next().copyTo( res ); UT_EQ ( "", res );  UT_TRUE( !tok2.hasNext() );
            UT_TRUE( !tok.hasNext() );
        }

    }

}
